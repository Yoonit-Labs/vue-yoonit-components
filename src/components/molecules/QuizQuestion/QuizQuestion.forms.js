export const answers = [
  {
    id: 'ans1',
    text: {
      ptBR: 'Cansaço'
    },
    weight: 3
  },
  {
    id: 'ans2',
    text: {
      ptBR: 'Corrimento Nasal (Coriza)'
    },
    weight: 3,
    status: true,
    override: false
  }
]
export const questionCheck = {
  id: 'covid-cl1',
  text: {
    ptBR: 'Marque os sintomas sentidos nas últimas horas'
  },
  visible: true,
  type: 'check',
  answers
}

export const questionAnswersNull = {
  id: 'covid-cl1',
  text: {
    ptBR: 'Marque os sintomas sentidos nas últimas horas'
  },
  visible: true,
  type: 'check',
  answers: []
}
export const questionRadio = {
  id: 'covid-cl1',
  text: {
    ptBR: 'auhsuhashuashuashuaus'
  },
  visible: true,
  type: 'radio',
  answers
}
