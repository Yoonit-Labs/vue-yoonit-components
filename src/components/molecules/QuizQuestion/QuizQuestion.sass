@import '~@yoonit/style'
@import '~@yoonit/style/molecules/YooQuizQuestion'
