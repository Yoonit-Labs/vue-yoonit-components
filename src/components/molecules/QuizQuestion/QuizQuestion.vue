<template lang="pug">
  YooGridLayout(
    rows="*"
    :cols="takeRows"
    :class="['yoo-quiz']"
  )
    p(
      row="1"
      col="1"
      :class="['yoo-quiz__question']"
    ) {{ question.text.enUS }}
      span(
        v-if="question.required"
      ) *

    CheckButton(
      v-for="(ans, index) in question.answers"
      :key="'radio-' + question.id + '-' + ans.id + '-' + index"
      :row="index+2"
      :col="1"
      :class="['yoo-quiz__card']"
      :textPosition="textPosition"
      :text="ans.text.enUS"
      :checked="doTakeStatus(index)"
      :card="true"
      @response='doTap($event, index, ans.override, question.type)'
    )

</template>

<script>

import YooGridLayout from '@/components/bosons/GridLayout/GridLayout.vue'
import CheckButton from '@/components/atoms/CheckButton/CheckButton.vue'
import PropsConfig from '@/components/molecules/QuizQuestion/QuizQuestion.config'

export default {
  name: 'YooQuizQuestion',
  props: {
    textPosition: {
      type: String,
      default: 'left',
      validator: value => PropsConfig.textPosition.options.includes(value)
    },
    question: {
      type: Object,
      required: true,
      validator: value => Object.prototype.hasOwnProperty.call(value, 'answers')
    }
  },
  components: {
    YooGridLayout,
    CheckButton
  },
  data: () => ({
    localAnswers: []
  }),
  mounted () {
    this.localAnswers = this.question.answers
  },
  computed: {
    takeRows () {
      return this.localAnswers
        .map(elm => 'auto')
        .join(',')
    }
  },
  methods: {
    doTap (event, answerId, override, type) {
      if (override || type === 'radio') {
        for (var answer in this.localAnswers) {
          this.localAnswers[answer].status = false
        }
      } else {
        for (var otherAnswer in this.localAnswers) {
          if (this.localAnswers[otherAnswer].override === true) {
            this.localAnswers[otherAnswer].status = false
          }
        }
      }
      this.$set(this.localAnswers[answerId], 'status', event)
      this.$emit('tapChoice', this.localAnswers)
    },
    doTakeStatus (ans) {
      return this.localAnswers && this.localAnswers[ans]
        ? this.localAnswers[ans].status
        : false
    }
  }
}
</script>

<style src="./QuizQuestion.sass" lang="sass" scoped></style>
