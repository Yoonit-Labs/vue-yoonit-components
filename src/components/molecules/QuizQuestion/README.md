# Yonnit Components

[Home](https://cyberlabs.ai/) | [Atoms](https://cyberlabs.ai/) | [Molecules](https://cyberlabs.ai/) | [Bosons](https://cyberlabs.ai/) | [Quarks](https://cyberlabs.ai/)

## YooQuizQuestions

The component renders a list of options to be selected.

## Usage

To use the component, simply invoke by passing the information in your props, question and answers.

### Input
`App.vue`
```vue
<template>
  <yoo-quiz-question
    textPosition="left"
    :question='{"id":"yoonit-cl1","text":{"enUS":"Are you enjoying using Yoonit Components"},"visible":true,"type":"radio","answers":[{"id":"ans1","text":{"enUS":"Yes"},"weight":3},{"id":"ans2","text":{"enUS":"No"},"weight":3,"status":true,"override":false}]}'
    :answers='[{"id":"ans1","text":{"enUS":"Yes"},"weight":3},{"id":"ans2","text":{"enUS":"No"},"weight":3,"status":true,"override":false}]'
  />
</template>
```

### Output

The output of the above code will be:

<img src="../../../../public/readme-img/quiz-questions.png" alt="YooQuizQuestion" width="800">

### Checkbox

<img src="../../../../public/readme-img/quiz-questions-checkbox.png" alt="YooQuizQuestion" width="800">

### Props

| Parameter | Type | Default | Valid values | Description | Required |
|-----------|------|------------------------|--------------|-------------|---------|
| **`textPosition`** | String | `left` | <ul><li>left</li></ul><ul><li>right</li></ul> | Position of text in the component. | false |
| **`question`** | Object | `''` | i.e {"id":"yoonit-cl1","text":{"enUS":"Are you enjoying using Yoonit Components"},"visible":true,"type":"radio","answers":[{"id":"ans1","text":{"enUS":"Yes"},"weight":3},{"id":"ans2","text":{"enUS":"No"},"weight":3,"status":true,"override":false}]} | Object with the questions that will be rendered. | true |

#### [**Next component**](../SegmentedBar/README.md) :arrow_forward:

#### :arrow_backward: [**Previous component**](../PostCard/README.md)
## To contribute and make it better

Clone the repo, change what you want and send PR.

Contributions are always welcome!

---

Code with ‚ù§ by the [**Cyberlabs AI**](https://cyberlabs.ai/) Front-End Team

