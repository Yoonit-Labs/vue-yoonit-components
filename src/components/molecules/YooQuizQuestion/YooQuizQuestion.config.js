module.exports = {
  textPosition: {
    options: [
      'left',
      'right'
    ]
  }
}
