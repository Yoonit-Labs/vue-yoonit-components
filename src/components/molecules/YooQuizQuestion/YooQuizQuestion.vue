<template lang="pug">
  YooGridLayout(
    rows="*"
    :cols="doSetRows(takeAnswers)"
    :class="['yoo-quiz']"
  )
    p(
      row="1"
      col="1"
      :class="['yoo-quiz__question']"
    ) {{ takeQuestion.text.ptBR }}
      span(
        v-if="takeQuestion.required"
      ) *

    YooSwitchCard(
      v-for="(ans, index) in takeQuestion.answers"
      :key="'radio-' + takeQuestion.id + '-' + ans.id + '-' + index"
      :row="index+2"
      :col="1"
      :class="['yoo-quiz__switch-card']"
      :textPosition="textPosition"
      :text="ans.text.ptBR"
      :checked="doTakeStatus(index)"
      @response='doTap($event, takeQuestion.id, index, ans.override, takeQuestion.type)'
    )

</template>

<script>

import YooGridLayout from '@/components/quarks/YooGridLayout/YooGridLayout.vue'
import YooSwitchCard from '@/components/molecules/YooSwitchCard/YooSwitchCard.vue'
import PropsConfig from '@/components/molecules/YooQuizQuestion/YooQuizQuestion.config'

export default {
  name: 'YooQuizQuestion',
  props: {
    textPosition: {
      type: String,
      default: 'right',
      validator: value => PropsConfig.textPosition.options.includes(value)
    },
    question: {
      type: Object,
      required: true
    },
    answers: {
      type: Array,
      required: true
    }
  },
  components: {
    YooGridLayout,
    YooSwitchCard
  },
  data: function () {
    return {
      localAnswers: []
    }
  },
  beforeCreate () {},
  created () {},
  beforeMount () {},
  mounted () {
    this.localAnswers = this.answers
  },
  beforeUpdate () {},
  updated () {},
  beforeDestroy () {},
  destroyed () {},
  computed: {
    takeQuestion () {
      return this.question
    },
    takeAnswers () {
      return this.question.answers || []
    }
  },
  methods: {
    doSetRows (answers) {
      return answers
        .map(elm => 'auto')
        .join(',')
    },
    doTap (event, questionId, answerId, override, type) {
      if ((override && event) || type === 'radio') {
        for (var answer in this.localAnswers) {
          this.$set(this.localAnswers[answer], 'status', false)
        }
      } else {
        for (var otherAnswer in this.localAnswers) {
          if (this.localAnswers[otherAnswer].override === true) {
            this.localAnswers[otherAnswer].status = false
          }
        }
      }

      this.$set(this.localAnswers[answerId], 'status', event)
      return this.$emit('tapChoice', this.localAnswers)
    },
    doTakeStatus (ans) {
      return this.localAnswers && this.localAnswers[ans]
        ? this.localAnswers[ans].status
        : false
    }
  },
  filters: {},
  watch: {}
}
</script>

<style src="./YooQuizQuestion.sass" lang="sass" scoped></style>
