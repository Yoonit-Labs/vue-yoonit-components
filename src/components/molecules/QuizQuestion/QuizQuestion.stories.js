/**
 * ██╗   ██╗ ██████╗  ██████╗ ███╗   ██╗██╗████████╗
 * ╚██╗ ██╔╝██╔═══██╗██╔═══██╗████╗  ██║██║╚══██╔══╝
 *  ╚████╔╝ ██║   ██║██║   ██║██╔██╗ ██║██║   ██║
 *   ╚██╔╝  ██║   ██║██║   ██║██║╚██╗██║██║   ██║
 *    ██║   ╚██████╔╝╚██████╔╝██║ ╚████║██║   ██║
 *    ╚═╝    ╚═════╝  ╚═════╝ ╚═╝  ╚═══╝╚═╝   ╚═╝
 *
 * https://yoonit.dev - about@yoonit.dev
 *
 * Vue Yoonit Components
 * VueJS Atomic Design System framework
 *
 * Vitória Costa, Tiago Brito, Fernando Junior, Sabrina Sampaio, Gabriel Mule, Gabriel Moraes, Gabriel Rizzo & Luigui Delyer @ 2020-2021
 */

import yooQuizQuestion from '@/components/molecules/QuizQuestion/QuizQuestion.vue'
import PropsConfig from '@/components/molecules/QuizQuestion/QuizQuestion.config'
import * as ConfigMocks from '@/components/molecules/QuizQuestion/QuizQuestion.forms'

var answers = [
  {
    id: 'ans1',
    text: {
      enUS: 'Yes'
    },
    weight: 3
  },
  {
    id: 'ans2',
    text: {
      enUS: 'No'
    },
    weight: 3,
    status: true
  }
]

export default {
  title: 'Yoonit/YooQuizQuestion',
  component: yooQuizQuestion,
  argTypes: {
    textPosition: {
      control: {
        type: 'select',
        options: PropsConfig.textPosition.options
      }
    }
  }
}

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { yooQuizQuestion },
  template:
    '<yoo-quiz-question v-bind="$props"></yoo-quiz-question>'
})

export const Custom = Template.bind({})
Custom.args = {
  question: ConfigMocks.questionRadio,
  answers: ConfigMocks.answers
}

export const CheckButton = Template.bind({})
CheckButton.args = {
  question: ConfigMocks.questionCheck,
  answers: answers
}
export const RadioButton = Template.bind({})
RadioButton.args = {
  question: ConfigMocks.questionRadio,
  answers: answers
}
