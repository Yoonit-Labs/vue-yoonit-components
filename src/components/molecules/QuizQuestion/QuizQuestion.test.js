import { mount } from '@vue/test-utils'
import PropsConfig from './QuizQuestion.config'
import YooQuizQuestion from './QuizQuestion.vue'
import * as ConfigMocks from '@/components/molecules/QuizQuestion/QuizQuestion.forms'

const SlotText = 'Default Slot Text'

const mountComponent = () => {
  return mount(YooQuizQuestion, {
    slots: { default: SlotText },
    propsData: {
      question: ConfigMocks.questionCheck
    }
  })
}

describe('YooQuizQuestion Component', () => {
  let wrapper
  beforeEach(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(0.123456789)
    wrapper = mountComponent()
  })

  afterEach(() => {
    jest.spyOn(global.Math, 'random').mockRestore()
  })

  it('Matches Snapshot', async () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Loads the Component HTML', () => {
    expect(wrapper.classes('yoo-quiz')).toBe(true)
  })

  describe('Props', () => {
    describe('textPosition', () => {
      it('Has a valid default value', () => {
        expect(PropsConfig.textPosition.options.includes(YooQuizQuestion.props.textPosition.default)).toBe(true)
      })
      PropsConfig.textPosition.options.forEach(textPosition => {
        it('Loads the Component HTML', async () => {
          await wrapper.setProps({ textPosition })
          expect(wrapper.find(`.yoo-checkbox__text--${textPosition}`).exists()).toBe(true)
        })
      })
    })

    describe('question required', () => {
      it('Does not `*` if question required', async () => {
        expect(wrapper.text()).not.toMatch('*')
      })
      it('Show `*` if question required', async () => {
        await wrapper.setProps({ question: { ...ConfigMocks.questionCheck, required: true } })
        expect(wrapper.text()).toMatch('*')
      })
    })

    describe('answers', () => {
      it('Includes class: .yoo-quiz__switch-card when array answers has lenght', async () => {
        expect(wrapper.find('.yoo-quiz__card').exists()).toBe(true)
      })
      it('Does not include class: .yoo-quiz__card when array answers does not has lenght', async () => {
        await wrapper.setProps({
          question: ConfigMocks.questionAnswersNull
        })
        expect(wrapper.find('.yoo-quiz__card').exists()).toBe(false)
      })
    })
  }) // describe Props

  describe('Events', () => {
    describe('Click', () => {
      it('Emits Click Event', async () => {
        await wrapper.find('.yoo-quiz__card').vm.$emit('response')
        expect(wrapper.emitted()).toHaveProperty('tapChoice')
      })
    })

    describe('Click', () => {
      it('Emits Click Event', async () => {
        await wrapper.setProps({ question: ConfigMocks.questionRadio })
        await wrapper.find('.yoo-quiz__card').vm.$emit('response')
        expect(wrapper.emitted()).toHaveProperty('tapChoice')
      })
    })
  }) // describe Events
})
