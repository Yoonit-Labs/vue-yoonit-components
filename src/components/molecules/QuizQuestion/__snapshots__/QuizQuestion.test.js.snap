// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooQuizQuestion Component Matches Snapshot 1`] = `
<div class="yoo-grid-layout yoo-quiz yoo-grid-layout--h-stretch yoo-grid-layout--v-stretch" style="grid-template-rows: auto; grid-template-columns: auto auto;">
  <p row="1" col="1" class="yoo-quiz__question" style="grid-row: 1; grid-column: 1;">Marque os sintomas sentidos nas últimas horas
    <!---->
  </p>
  <div class="yoo-checkbox yoo-quiz__card yoo-checkbox__card" row="2" col="1" style="grid-row: 2; grid-column: 1;"><input type="checkbox" id="ycid-3456789"><label class="yoo-checkbox__check fas fa-check yoo-checkbox--medium p__t--l p__r--l p__b--l p__l--l yoo-checkbox__text--left" for="ycid-3456789">
      <p class="yoo-checkbox__text yoo-checkbox--medium">Cansaço</p>
    </label></div>
  <div class="yoo-checkbox yoo-quiz__card yoo-checkbox__card yoo-checkbox__card--active" row="3" col="1" style="grid-row: 3; grid-column: 1;"><input type="checkbox" id="ycid-3456789"><label class="yoo-checkbox__check fas fa-check yoo-checkbox--medium p__t--l p__r--l p__b--l p__l--l yoo-checkbox__check--card-active yoo-checkbox__text--left" for="ycid-3456789">
      <p class="yoo-checkbox__text yoo-checkbox--medium">Corrimento Nasal (Coriza)</p>
    </label></div>
</div>
`;
